### ISSUE TITLE
Prepend breaks module "implementer" search

### ISSUE THREAD

**Issue by headius** [jruby/jruby#2864]
This code should work and it doesn't. This is the cause of the Rails master "each_with_index" bug when running tests:

```ruby
module A
  def foo
    p 'ok'
  end
end

class X
  include A
end

class Y < X
end

module B
end

class Y
  prepend B
end

module A
  prepend B
end

Y.new.foo
```

I believe the problem is that our "findImplementer" logic for module methods ends up jumping above Enumerable in the search, due to the same module appearing in the hierarchy twice (due to lazy prepends).

**Comment by headius** [comment]
Ok, I think I'm wrong. In monitoring a method search for the code above, we *do* walk the hierarchy correctly, but when it gets to the `A` module it only searches its direct method table. Because we've prepended something into A, that table is empty.

My original fix for this was to make IncludedModuleWrapper keep searching prepended items until it finds a non-prepended, but that didn't seem to fix it.

**Comment by headius** [comment]
The problem here was that I broke my prepend-searching fix.

In 7ac2331d0567943fbea023274a6d9a23b9dd7d18, I fixed prepends-in-a-module by searching the module's ancestors. However I searched too deeply, going into includes above the module that normally are not searched (they are copied into target hierarchy at include time).

Then in c725c6984de7e9b1b4132d9fb8cd10739b7050c7 I fixed the search to not go so deep...but I broke it by making it start at `this` – the IncludedModuleWrapper – rather than at `origin` – the actual module. That resulted in it going back to old behavior.

The new fix, in b9473af, restores the proper starting point and terminates the search by looking for methodLocation (which indicates the module under which prepends should go). The final search covers the module itself, and all is well.
